Hosting API review feedback # 1

Praburaj opened the pull request:
Addresses:
https://github.com/aspnet/Hosting/issues/164
https://github.com/aspnet/Hosting/issues/163
https://github.com/aspnet/Hosting/issues/155
https://github.com/aspnet/Hosting/issues/156
https://github.com/aspnet/Hosting/issues/154

@loudej for https://github.com/aspnet/Hosting/issues/164

msotclas commented:
Hi __@Praburaj__, I'm your friendly neighborhood Microsoft Open Technologies, Inc. Pull Request Bot (You can call me MSOTBOT). Thanks for your contribution!
    <p>
        It looks like you're a Microsoft contributor (Praburaj Thiagarajan). If you're full-time, we DON'T require a Contribution License Agreement. If you are a vendor, please DO sign the electronic Contribution License Agreement. It will take 2 minutes and there's no faxing! https://cla2.msopentech.com.
    </p>

TTYL, MSOTBOT;

davidfowl commented on src/Microsoft.AspNet.Hosting/Builder/HttpContextFactory.cs:
```diff
@@ -9,11 +9,10 @@ namespace Microsoft.AspNet.Hosting.Builder
-            var featureObject = serverContext as IFeatureCollection ?? new FeatureObject(serverContext);
-            var featureCollection = new FeatureCollection(featureObject);
-            var httpContext = new DefaultHttpContext(featureCollection);
+            var featureObject = featureCollection ?? new FeatureObject(null);
```
Just use the passed in feature collection. This should never be null.

davidfowl replied:
Just use the passed in feature collection. This should never be null.

davidfowl replied:
Just use the passed in feature collection. This should never be null.

davidfowl replied:
Just use the passed in feature collection. This should never be null.

davidfowl replied:
Just use the passed in feature collection. This should never be null.

davidfowl replied:
Just use the passed in feature collection. This should never be null.

davidfowl replied:
Github fail

Praburaj replied:
okay.

Eilon commented on src/Microsoft.AspNet.Hosting/HostingContext.cs:
```diff
@@ -11,7 +10,7 @@ namespace Microsoft.AspNet.Hosting
     public class HostingContext
     {
-        public IServiceProvider Services { get; set; }
+        public IApplicationLifetime ApplicationLifeTime { get; set; }
```
I think `Lifetime` is one word.

Praburaj replied:
Will fix this.

Eilon commented on src/Microsoft.AspNet.Hosting/HostingEngine.cs:
```diff
@@ -105,9 +106,16 @@ private void EnsureApplicationStartup(HostingContext context)
+
+            if (context.ApplicationStartup == null)
+            {
+                throw new Exception(diagnosticMessages.Aggregate("TODO: web application entrypoint not found message", (a, b) => a + "\r\n" + b));
```
Use appropriate exception type. This is probably an ArgumentException, with `context` being the invalid argument.

Also, no more "TODO" in the error message - just put whatever the error message should be.

Praburaj replied:
done.

Eilon commented on src/Microsoft.AspNet.Hosting/Startup/StartupLoader.cs:
```diff
@@ -94,9 +90,9 @@ private object Invoke(MethodInfo methodInfo, object instance, IApplicationBuilde
+            if (string.IsNullOrEmpty(applicationName))
             {
-                return _next.LoadStartup(applicationName, environmentName, diagnosticMessages);
+                throw new ArgumentNullException("applicationName");
```
Wrong exception type. This is thrown for null or empty, yet the exception is for null only. Use a regular ArgException with a message that says can't be null or empty (copy whatever message we used in WSR).

Eilon commented on src/Microsoft.AspNet.TestHost/TestServer.cs:
```diff
@@ -24,20 +22,21 @@ public class TestServer : IServerFactory, IDisposable
         public TestServer(IConfiguration config, IServiceProvider serviceProvider, Action<IApplicationBuilder> appStartup)
         {
             var appEnv = serviceProvider.GetRequiredService<IApplicationEnvironment>();
+            var applicationLifeTime = serviceProvider.GetRequiredService<IApplicationLifetime>();
```
See earlier comment about LifeTime --> Lifetime.

Praburaj replied:
Will fix this.

Eilon commented:
Overall looks good, just some further cleanup needed.

Tratcher commented on src/Microsoft.AspNet.Hosting/HostingContext.cs:
```diff
@@ -21,7 +20,7 @@ public class HostingContext
         public RequestDelegate ApplicationDelegate { get; set; }
 
-        public string ServerName { get; set; }
+        public string ServerFactoryAssembly { get; set; }
```
That this is the name of an assembly is an implementation detail of the default IServerFactory. How about ServerFactoryLocation?

Praburaj replied:
Sure. That makes sense. I can fix that.

Tratcher commented on src/Microsoft.AspNet.Hosting/HostingEngine.cs:
```diff
@@ -39,16 +40,16 @@ public IDisposable Start(HostingContext context)
             EnsureApplicationDelegate(context);
 
-            var applicationLifetime = (ApplicationLifetime)context.Services.GetRequiredService<IApplicationLifetime>();
+            var applicationLifetime = (ApplicationLifetime)context.ApplicationLifeTime;
```
This cast makes it impossible to replace this service. Maybe the interface needs to be re-worked.

Praburaj replied:
Yeah. But the cast is not introduced with this change. So I would keep that as a separate work item rather than mixing it with API review.

Praburaj replied:
Actually I think I can fix it now. `IApplicationLifetime` to include these method names.

Praburaj commented:
I addressed all comments except https://github.com/aspnet/Hosting/pull/167#discussion_r25447213. I'm filing a work item and fix that separately. Merging this change in.

Eilon commented on src/Microsoft.AspNet.Hosting/Startup/StartupLoader.cs:
```diff
@@ -94,9 +90,9 @@ private object Invoke(MethodInfo methodInfo, object instance, IApplicationBuilde
+            if (string.IsNullOrEmpty(applicationName))
             {
-                return _next.LoadStartup(applicationName, environmentName, diagnosticMessages);
+                throw new ArgumentException("applicationName");
```
Wrong parameter to this exception. ArgException takes either "message" or "message, paramName". Use this message: https://github.com/aspnet/Mvc/blob/f6e701b2b2d01ae068b144416aeac080c623bc17/src/Microsoft.AspNet.Mvc.Core/Resources.resx#L205-L207

Praburaj replied:
ah..fixed.
